File: go/src/queue.go
package ballclock

type Queue struct {
	balls      []int
	currentPos int
}

func NewQueue(cap int) *Queue {
	q := new(Queue)
	q.balls = make([]int, cap, cap)

	for i := 0; i < cap; i++ {
		q.AddBall(i + 1)
	}

	return q
}

func (q *Queue) GetBall() int {
	b := q.balls[0]

	tmp := make([]int, len(q.balls), cap(q.balls))
	copy(tmp, q.balls[1:])
	q.balls = tmp
	q.currentPos -= 1

	return b
}

func (q *Queue) AddBall(b int) {
	q.balls[q.currentPos] = b
	q.currentPos += 1
}

File: c#/BallClock/Queue.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace BallClock
{
    public class Queue
    {
        public List<int> Balls { get; private set; }

        public Queue()
        {
            this.Balls = new List<int>();
        }

        public void AddBall(int ball)
        {
            this.Balls.Add(ball);
        }

        public int this[int index]
        {
            get
            {
                return this.Balls[index];
            }
        }

        public int GetBall()
        {
            var ball = this.Balls.First();
            this.Balls.RemoveAt(0);
            return ball;
        }

        public int Length
        {
            get
            {
                return this.Balls.Count();
            }
        }

        public int First()
        {
            return this.Balls.First();
        }

        public int Last()
        {
            return this.Balls.Last();
        }
    }
}
